machine-runner/src/test-utils/mock-runner.ts

```diff
@@ -0,0 +1,197 @@
+  ]
+  export const make = () => {
+    const data = {
+      isDelaying: false as boolean,
+      delayedCommands: [] as Pair[],
+    }
+
+    const make = (): Pair => {
+      let finished = false
+      const pair: Pair = [undefined as any, undefined as any]
+      pair[0] = new Promise<void>((resolve, reject) => {
+        pair[1] = {
+          resolve,
+          reject,
+          isFinished: () => finished,
+        }
+      }).finally(() => (finished = true))
+
+      if (data.isDelaying) {
+        data.delayedCommands.push(pair)
+      } else {
+        pair[1].resolve()
+      }
+      return pair
+    }
+
+    const toggle = async (
+      delayControl: { delaying: true } | { delaying: false; reject?: boolean },
+    ): Promise<void> => {
+      data.isDelaying = delayControl.delaying
+      if (delayControl.delaying) return
+      await Promise.all(
+        data.delayedCommands.map(([promise, control]) => {
+          if (delayControl.reject) {
+            control.reject()
+          } else {
+            control.resolve()
+          }
+          // eslint-disable-next-line @typescript-eslint/no-empty-function
+          return promise.catch(() => {})
+        }),
+      )
+      data.delayedCommands = []
+    }
+
+    return {
+      make,
+      toggle,
+    }
+  }
+}
+
+export namespace Subscription {
+  export const make = <RegisteredEventsFactoriesTuple extends MachineEvent.Factory.Any[]>() => {
+    type Subscribe = SubscribeFn<RegisteredEventsFactoriesTuple>
+    type Cb = Parameters<Subscribe>[0]
+    const cancel = () => {
+      if (data.cb === null) throw new Error('not subscribed')
+      data.cb = null
+    }
+    const subscribe: Subscribe = (onEvent) => {
+      if (data.cb !== null) throw new Error('already subscribed')
+      data.cb = onEvent
+      return data.cancel
+    }
+
+    const data = {
+      cb: null as null | Cb,
+      cancel,
+      subscribe,
+    }
+    return data
+  }
+}
+
+export const createMockMachineRunner = <
+  SwarmProtocolName extends string,
+  MachineName extends string,
+  RegisteredEventsFactoriesTuple extends MachineEvent.Factory.Any[],
+  Payload,
+>(
+  factory: StateFactory<
+    SwarmProtocolName,
+    MachineName,
+    RegisteredEventsFactoriesTuple,
+    string,
+    Payload,
+    object
+  >,
+  payload: Payload,
+) => {
+  const delayer = PromiseDelay.make()
+  const sub = Subscription.make<RegisteredEventsFactoriesTuple>()
+  const persisted: MachineEvent.Any[] = []
+
+  const feed = (
+    ev: MachineEvent.Factory.ReduceToEvent<RegisteredEventsFactoriesTuple>[],
+    props?: { caughtUp: true },
+  ) => {
+    const caughtUp = props === undefined || props.caughtUp === true
+    if (sub.cb === null) throw new Error('not subscribed')
+    return sub.cb({
+      type: MsgType.events,
+      caughtUp,
+      events: ev.map((payload) => ({
+        meta: mockMeta(),
+        payload,
+      })),
+    })
+  }
+
+  const timeTravel = () => {
```

Comment:
not sure whether we should offer this one: the mock runner may not be started from the real initial state, and then the user needs to take care to feed the full history afterwards — I’d say we leave this out for now and observe what functionality we’ll need

Reply:
Will do.

Reply:
As a note, we will reuse this mock-runner for our own internal unit tests too